Add support for custom backend

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95072

Fixes https://github.com/pytorch/pytorch/issues/92344

A custom backend can be specified by passing in a string with format `"<device_type1>:<backend_name>,<device_type2>:<backend_name>"`, e.g. `"cpu:gloo,cuda:custom_backend"`.

Differential Revision: [D43630050](https://our.internmc.facebook.com/intern/diff/D43630050)

cc @ezyang @gchanan @zou3519

H-Huang committed:
Add support for custom backend

[ghstack-poisoned]

H-Huang committed:
Update on "Add support for custom backend"




[ghstack-poisoned]

Zha0q1 commented:
Thanks @H-Huang for starting this PR!

> A custom backend can be specified by passing ina string with format "<device_type1>:<backend_name>,<device_type2>:<backend_name>", e.g. "cpu:gloo,cuda:custom_backend"
Will dist.init_process_group(backend='custom') work? If so, will all devices be mapped to "custom"? Asking because existing frameworks and training scripts have been using this semantic.

H-Huang commented:
Hi @Zha0q1,

> Will dist.init_process_group(backend='custom') work? If so, will all devices be mapped to "custom"? Asking because existing frameworks and training scripts have been using this semantic.

Yes, that will continue to work as it does in PT1.0. `dist.init_process_group(backend='custom')` will be equivalent to config `cpu:custom,cuda:custom`

H-Huang committed:
Update on "Add support for custom backend"


Fixes https://github.com/pytorch/pytorch/issues/92344

A custom backend can be specified by passing in a string with format `"<device_type1>:<backend_name>,<device_type2>:<backend_name>"`, e.g. `"cpu:gloo,cuda:custom_backend"`.

[ghstack-poisoned]

Zha0q1 commented:
Hi @H-Huang, we will start testing this PR from our side too, thank you so much for making this happen! Just wanted to confirm, is this PR pending any additional commits, or we can already take it to testing?

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43630050).

H-Huang commented:
@Zha0q1 you can test this PR! It is under review, but I don't expect any major changes

Zha0q1 commented:
I have built and tested the latest commit ede42e6a24b95dca7643c8ef8ec053808b2d3b93
Custom backend is correctly handled now -- 
`backend='custom'` creates only one backend to handle both cpu and cuda tensors
`backend=cpu:gloo,cuda:custom` creates a gloo backend for cpu and a custom backend for cuda

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -243,13 +243,27 @@ def __init__(self, backend: Union[str, Backend]):
+                    raise ValueError(f"Duplicate device type {device} in backend string: {backend}")
+                self.device_backend_map[device] = Backend(backend)
+
+            # ensure cpu and cuda are specified
+            required_devices = ["cpu", "cuda"]
+            for device in required_devices:
+                if device not in self.device_backend_map:
+                    raise ValueError(f"Device type {device} is not specified in backend string: {backend}")
```
I wonder if this check would be too restrictive for non-CUDA, non-CPU use cases?

H-Huang replied:
Hmm yeah, I suppose that if user doesn't specify "cpu" or "cuda" then process group won't error out unless explicitly called with a "cpu" or "cuda" tensor, so we don't need this. I can remove this check

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1088,15 +1104,15 @@ def _new_process_group_helper(
-        # only create single backend pg when backend is set to gloo, nccl, mpi, and ucc
-        if backend in [Backend.GLOO, Backend.NCCL, Backend.UCC, Backend.MPI]:
+        # make a single backend when all get_device_backend_map values are the same
+        if set(backend_config.get_device_backend_map().values()) == {backend_str}:
```
Maybe:
`if len(...) == 1`
to make it simpler?

H-Huang replied:
Thanks! updated

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -243,13 +243,27 @@ def __init__(self, backend: Union[str, Backend]):
-            assert device in self.device_backend_map
+            # make sure the backend string is in the correct format:
+            # "{device_type1}:{backend1},{device_type2}:{backend2}"
+            pattern = r'^[a-zA-Z0-9]+:[a-zA-Z0-9]+(,[a-zA-Z0-9]+:[a-zA-Z0-9]+)*$'
```
Maybe more comments for this syntax if you can?

An idea for avoiding using `re` at all: 
just go ahead to `split` the string by `,` and then split by `:` (as your code below does). If the returned split list's length `!= 2`, that means either `:` is not found or there are too many `:`.

H-Huang replied:
Yeah I agree the check can be done that way. I think regex might be a little more manageable and result in less if conditions. It also limits the inputs to alphanumeric characters. For example if user does something accidentally like `cpu;:gloo` then it would error with regex (due to the `;`) but we would need another case to check for special characters. If you want to checks for each `:` pair I can add that instead, lmk!

Updated the comments as well.

kwen2501 replied:
> It also limits the inputs to alphanumeric characters

Then something like "my_awesome_backend" would not work?

H-Huang replied:
hmm thats right, do you think it is necessary?

H-Huang committed:
Update on "Add support for custom backend"


Fixes https://github.com/pytorch/pytorch/issues/92344

A custom backend can be specified by passing in a string with format `"<device_type1>:<backend_name>,<device_type2>:<backend_name>"`, e.g. `"cpu:gloo,cuda:custom_backend"`.

Differential Revision: [D43630050](https://our.internmc.facebook.com/intern/diff/D43630050)

[ghstack-poisoned]

H-Huang committed:
Update on "Add support for custom backend"


Fixes https://github.com/pytorch/pytorch/issues/92344

A custom backend can be specified by passing in a string with format `"<device_type1>:<backend_name>,<device_type2>:<backend_name>"`, e.g. `"cpu:gloo,cuda:custom_backend"`.

Differential Revision: [D43630050](https://our.internmc.facebook.com/intern/diff/D43630050)

[ghstack-poisoned]

kwen2501 reviewed (approved):
Thanks for the update. LGTM!

H-Huang commented:
@pytorchbot merge

atalman commented:
cc @H-Huang 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176
